### ISSUE TITLE
Overall plan

### ISSUE THREAD

**Issue by richfitz** [mrc-ide/hipercow#8]
# package structure

Having a mega package is not ideal, because:

* we're going to have to do a lot of internal namespacing (e.g., `windows_valid_clusters()` vs `valid_clusters()`) and we expect that there won't be much code shared between the different backends/drivers
* the code that we end up running on the cluster will have a ton of extra dependencies (e.g., for the http clients, ssh clients, jsonlite, xml2, etc). This is not ideal as it makes the bootstrap harder to support and upgrades more error prone as they are subject to windows file locking rules
* if we do things in one package, we don't know that our abstraction is ok, and it's hard to rewrite/add a driver

Having a split package is not ideal, because:

* we end up with tight coupling of component packages, which makes it not obvious where to add code, and where to document things - this might be somewhat solvable by using a monorepo though.
* the users have to think about what they are installing and using too much - ideally we have them only interacting with a single package.
* we need to do some complex machinery for registering different "backends"

OTOH, there *is* a complexity for having some sort of support for different backends, so some of this is unavoidable, and it's just a case of working out the most sustainable place to hide it.

In order to have the users interact only with `hermod` we might develop:

* `hermod.core` - the bit that runs on the cluster - create a root, create a generic task, run a task etc.
* `hermod.windows` - the windows driver
* `hermod.linux` - the linux driver
* `hermod` - a fairly lightweight package that imports the others and contains the docs

Lay these out in the repo as

* `core/`
* `windows/`
* `linux/`
* `hermod/` - build the pkgdown from here alone

everything installs from universe, users get all drivers all the time.

# Example interactions

Consider just the simplest use case here - equivalent to the old `obj$enqueue(sessionInfo())` "hello world" that we have used in the past, and different user interactions.

## Windows cluster user

Collect credentials, this is done only once per DIDE password change:

```r
hermod::hermod_dide_credentials()
```

* prompts for keychain password, DIDE username, DIDE password
* validates the password by logging in
* saves the username and password in the keychain

Initialise a repo somewhere, telling it we'll be using windows as the driver

```r
hermod::hermod_init(".", driver = "windows")
```

Alternatively we might have done

```
hermod::hermod_init(".")
hermod::hermod_configure("windows")
```

which would be easier to pass through different arguments.

On the old tools the user would do:
```
share <- didehpc::path_mapping("malaria", "M:", "//fi--didef3.dide.ic.ac.uk/malaria", "M:")
config <- didehpc::didehpc_config(shares = share)
```

Here they might write:

```
share <- hermod::hermod_remote_path(
  path_local = "M:",
  path_remote = "//fi--didef3.dide.ic.ac.uk/malaria", 
  drive_remote = "M:")
hermod::hermod_configure("windows", shares = share)
```

We should consider these names very carefully so that they are appropriate for both linux and windows hosts and clients.

Provision the packages for the cluster (see later for more on how this might vary)

```r
hermod::hermod_provision()
```

This will run a cluster job that provisions the library at 
`hermod/lib/<platform>/<version>/`, then stream back logs from the job. Because we'll be using the windows backend here, that's just the case of monitoring a file at a known location, but this would be abstracted by the general monitoring commands that we'll describe below.  The process here really will be easier to understand after understanding the overall job submission, and in the first versions of implementation we will skip this by using a bootstrapped version installed in a temporary directory that we wire up to work with the `R_LIBS_USER` environment variable directly in the batch file.

### Queueing tasks

Enqueue a job

```r
id <- hermod::hermod_enqueue(sessionInfo())
```

This will:

* create a new task; we'll sort out some sort of `rlang`-based handling of quoted/unquoted expressions, which will simplify some of the existing handling.
* write out a windows-specific batch file within the task directory.
* run the submitter; use the web client here to pass in the path to this created batch file
* record the DIDE id of the submitted job, which we need in order to query back to queue for some information.

Previously we returned a "task" object that let people write

```
t$status()
```

but here we're tending towards

```
hermod::hermod_task_status(id)
```

which is what we do in `rrq` and is ok in practice.  We could provide an oopy interaface later if people prefer it perhaps?

Other job submission approaches we might explore

```r
hermod::hermod_enqueue_call("orderly::orderly_run", "my-report", list(a = 1))
hermod::hermod_enqueue(sqrt({{x}}))
hermod::hermod_enqueue(sqrt({{x}}), resources = list(cores = 10))
```

### Task status

Enquire about the task status

```r
hermod::hermod_task_status(id)
```

This could operate two ways depending on the backend:

* just check the status within the local copies of the files. If they are not submitted or if they are completed there is no point interacting with the remote status
* if submitted/running we might want to check the true status; this requires hitting the web api which will be a little slower, and ideally we can do this with a set of ids so that this function can be vectorised

Ways that things can get out sync:

* failure to map/mount a network share, nothing gets updated, job stays pending (i.e., submitted)
  * share is being hammered (e.g., home directory)
  * user provides incorrect share information and the mount fails
  * node is being hammered and network can't connect
* some hermod bug, so that job starts, hands off to the R script perhaps, but fails before the status updates, job stays pending
* loss of node, loss of node network after start, job stays running
* segfault of task, or OOM, job stays running (could mitigate by writing a status file in the batch file as a fallback though)
* out of disk space, can end up in almost any stale status, generally terrible and especially when multiple jobs are running at once

### Task logs

As the task runs, we want to be able to fetch the logs:

```r
hermod::hermod_task_log(id)
```

* For the windows version we can just read the file present in the directory
* We might offer a "follow" argument here, we use that in conan already

we might need to hit the DIDE log, which wraps the above:

```r
hermod::hermod_task_log(id, wrapper = TRUE)
```

### Get result

Once the task has finished we do

```r
hermod::hermod_task_result(id)
```

* Simply read the rds file that is already there in the directory

### Task deletion

On deletion, cancel and unsubmit any jobs that are still running/queued. Do this in *reverse* order, which is easy because the DIDE id is an integer. Do this in bulk, Wes will get an API call up for us.

Higher level function for common issues:

* "Delete everything currently running"
* "Delete everything submitted since/before X"
* "Delete from a bundle"

Need to think about dependent jobs here too of course.

Cancelling and deletion might be separate sometimes, but usually not.

Something to blow away "everything":

* the tasks?
* the library?

### Linux cluster user, from their laptop

This is not going to be enough

```
hermod::hermod_init(".", driver = "linux")
```

Because we *must* provide a remote working directory!

```
hermod::hermod_init(".", driver = "linux", remote_path = "~/malaria/sites")
```

```
hermod::hermod_init(".")
hermod::hermod_configure("linux", remote_path = "~/malaria/sites")
```

additional shares would need mapping again, but that will depend on how easy it is to mount them, and where on the node they get mounted.

Provisioning would likely be about the same.

### Queueing tasks

Enqueue a job

```r
id <- hermod::hermod_enqueue(sessionInfo())
```

This will:

* create a new task; we'll sort out some sort of `rlang`-based handling of quoted/unquoted expressions, which will simplify some of the existing handling.
* write out a linux-specific bash script and scp that to the task directory, probably it never lives locally
* scp the task expression to the cluster
* exec the qsub over ssh
* record the pbs id of the submitted job, which we need in order to query back to queue for some information.

### Task status

Enquire about the task status

```r
hermod::hermod_task_status(id)
```

* directly - ssh in and run `qstat`, save the found status locally
* don't check terminal status over ssh
* if qstat is slow, consider doing an existance check on a status file in the directory first to see if its worthwhile

### Task logs

As the task runs, we want to be able to fetch the logs:

```r
hermod::hermod_task_log(id)
```

* ssh exec cat the logfile, or tail after some known length
* copy over (or cache in memory) the log so far so that we increment
* once the task has finished we save the log locally so that this is nice and fast.

we might need to hit the PBS log, which wraps the above:

```r
hermod::hermod_task_log(id, wrapper = TRUE)
```

we'll work out from the tool how best to do that, perhaps `qlog`

### Get result

Once the task has finished we do

```r
hermod::hermod_task_result(id)
```

* scp file locally, then read it.

### Linux cluster user, from an ssh session

No arguments to init, because we know where we are. We could probably just let the driver be "linux" and sniff the hostname really to work out if they have ssh'd in or not!

```
hermod::hermod_init(".", driver = "linux")
```

Enqueue a job

```r
id <- hermod::hermod_enqueue(sessionInfo())
```

This will:

* create a new task, as above
* write out a linux-specific bash script, locally
* exec qsub via system2
* record the pbs id of the submitted job, which we need in order to query back to queue for some information.

# Stocktake

* How many tasks do I have? how old are they? how much space do they take up? What can I delete?

# Provisioning

* respond automatically to the files the user provides so that we do the right thing for provisining (renv vs pkgdepends, etc)
* cope with mutated libraries - how do we know if things are up to date
* skip provisioning sometimes
* need to cope with .tar.gz files, private repos, both of which are hard to get right

# Compacting tasks

We end up hitting the disk a lot in a way that seems pretty slow in practice, so it would be nice to have a better approach for doing this. One solution would be to write out the pieces of completed jobs into some sort of single-file structure that can be efficiently read - something like arrow/parquet perhaps? Any terminal job could be written in that way, though things like redirects need some work to think about.

Possibly sharding the directories (like in orderly2) will help here.

Optionally, trim objects out of the queued expressions to save space (lossy but commonly useful)

# Other weird stuff

## Packages loaded are not necessarily the packages installed

Provisioning will install a bunch of packages, that's great.

When a task runs it might assume a *subset* of these are attached, and that's harder to specify.

Consider the function:

```
f <- function(a) {
  a %>% sqrt
}
```

if we enqueue this function with

```
hermod_enqueue(f(10))
```

it will fail because `%>%` is not found! We need to somehow tell it to attach `dplyr` before running:

```
hermod_enqueue(f(10), packages = "dplyr")
```

This will get a bit boring for users, so we might also want to do something to allow people to say:

```
hermod_set_default_packages(c("dplyr", "tidyr"))
```

and then have those packages loaded at the start of any job.

Another option would be to set `packages = "auto"` in the submission (and possibly in the `set_default_packages`) which would try and load _all_ packages in your environment on the cluster, though we'd exclude hermod from this. Or we'd look at the *intersection* of your loaded packages and the packages found in the remote system perhaps. We should follow `search()` in terms of loading packages here.

# Requesting resources

The simplest approach could be:

```
hermod_submit(f(), resources = list(cores = 10))
hermod_submit(f(), resources = hermod::hermod_resources(cores = 10))
```

that could get a bit tedious to write each time so the *users* could say:

```
hermod_set_resource("parallel", cores = 10)
```

and then submit with

```
hermod_submit(f(), resources = "parallel")
```

with a special "default" name that would be use by default. We'd want to provide

```
hermod_list_resource()
```

which would list the various set types, and store this somewhere in the root.

* `cores` (meaning all on one node)
* `nodes` (with 1 being equal to "wholenode")
* `memory`
* `gpu`
* `kill_me_if_i_use_too_much_ram`
* `max_walltime`
* `cluster` (if we end up with more than one; still no plan to support old cluster though; we could potentially automatically throw things at the "most free" cluster, with some periodic check)

We should find out if the scheduler kills jobs that go out of memory?

# Dependencies among jobs

Alex implemented with the old tools, this exists in the underlying portal and hpc kit

```
hermod_submit(f(), depends_on = id)
```

The id here is the *hermod* id, can be a vector (in which case it's all ids). We remap these to DIDE ids if they're not already completed at the point of submission and let the queue look after it.

Jobs can become impossible in this situation, so consider that.

Jobs can be reqeueued after failure which would change their ID so consider that.

# Notifications

Different notification backends, needed here:

* email
* slack
* teams
* going "bing"

This could be done either by the hpc kit or by the tooling.

```
hermod_set_notifier("teams") # enable the notifier, collect credentials etc
hermod_set_notifier("email", address = c("r.fitzjohn@imperial.ac.uk", "w.hinsley@imperial.ac.uk"))
hermod_enqueue(f(), notify = "teams")
```

# List of existing grievances

Wes has a list!

> I am going to submit 24 single core jobs but I want these to land on different nodes

perhaps if we add some memory resource request (e.g., in GB) and default that to the minimum average amount of ram-per-core over the cluster the scheduler can take care of the rest.

```
hermod_submit(f(), resources = list(cores = 1, memory = 8))
```

> Job has failed with some weird code that only wes knows

return this code to the user somehow - it must be gettable from the HPC kit if it renders in the GUI. We could look these up, provide a table, or provide a link for the user to look them up. A lot of time is spent for people to work out that their jobs have a null pointer exception.

> My job failed and I am not sure what I need to provide in order to get help

```r
hermod::hermod_rescue(id)
```

Collect up all of the relevant config, logs, paths, job id, dide id, package that up as a little zip file and throw it at the cluster channel or something

**Comment by richfitz** [comment]
all moved into tickets now
